File: doomsday.c
#include <stdio.h>
#include <stdlib.h>

struct element{
    char symbol;
    int x;
    int y;
    int time;
};

struct node {
    struct element *elem;
    struct node *next;
};

struct queue {
    struct node *first, *last;
};

void enqueue(struct queue *q, char c, int x, int y, int t) {
    struct node *new_node = malloc(sizeof(struct node));
    if (!new_node) {
        perror("Error: ");
        return;
    }
    struct element *elem = malloc(sizeof(struct element));
    if(!elem) {
        perror("Error: ");
        return;
    }
    new_node->elem = elem;

    new_node->elem->symbol = c;
    new_node->elem->x = x;
    new_node->elem->y = y;
    new_node->elem->time = t;
    new_node->next = NULL;

    if (q->last)
        q->last->next = new_node;
    else
        q->first = new_node;
    q->last = new_node;
}

struct element *dequeue(struct queue *q) {
    if (q->first == NULL) return NULL; //-1;

    struct element *ret = q->first->elem;

    struct node *tmp = q->first;

    q->first = q->first->next;
    if (!q->first) {
        q->last = NULL;
    }
    tmp->next = NULL;
    free(tmp);

    return ret;
}

int main(int argc, char **argv) 
{ 
    if (argc != 2) 
    {	
        printf("Usage: ./doomsday text_file\n");
        return -1;
    }

    FILE *fp;	
    int i=0, j=0, k=0, l=0, time=-1, t=0, doomsday=0;
    char c, buff[1000][1000],positive_grid[1000][1000], negative_grid[1000][1000];
    struct queue *q;
    q = malloc(sizeof(struct queue));
    q->first = q->last = NULL;

    fp = fopen(argv[1], "r");
    if (fp == NULL)
    {
        perror("Error: ");
        return -1;
    }

    c = getc(fp) ;
    while (c != EOF)
    {
        buff[i][j] = c;

        if (c == '+' || c == '-'){
            enqueue(q, c, i, j, t);
            //printf("%c %d %d %d\n", c,i,j,t);
            if(c == '+')
                positive_grid[i][j] = '+';
            else if(c == '-')
                negative_grid[i][j] = '-';
        }

        j++;
        c = getc(fp);

        if(j == 1000)
        {
            i++;
            k=i;
            l=j;
            j=0;
        }

        if(c == '\n')
        {
            i++;
            k=i;
            l=j;
            j=0;
            while(c == '\n')
                c = getc(fp);
        }
    }	

    /*	for(i=0; i<k; i++)
        {
        for(j=0; j<l; j++)
        {	
        printf("%c", buff[i][j]);e
        }
        printf("\n");
        } */

    int up=0, down=0, right=0, left=0;

    struct element *elem;
    elem = dequeue(q);

    while(elem)
    {
        c = elem->symbol;
        i = elem->x;
        j = elem->y;
        t = elem->time;

        if((doomsday) && (time < t)){
            //printf("Broke loop\n");
            break;
        }

        if(((c == '-') && (buff[i][j] == '+')) || ((c == '+') && (buff[i][j] == '-')))
        {
            buff[i][j] = '*';
            if (!doomsday) {
                doomsday = 1; 	
                time = t;
            }
        }
        else if (buff[i][j] != '*') {
            if ((time == -1) || (t <= time)) {
                buff[i][j] = c;

                up = i - 1;
                down = i + 1;
                right = j + 1;
                left = j - 1;

                if ((up >= 0) && (buff[up][j] != 'X') && (buff[up][j] != c)) {
                    if ((c == '+') && (positive_grid[up][j] != '+')) {
			positive_grid[up][j] = '+';	
                        enqueue(q, c, up, j, t+1);
                    }
                    if ((c == '-') && (negative_grid[up][j] != '-')) {
                        negative_grid[up][j] = '-';
                        enqueue(q, c, up, j, t+1);
                    }
                    //printf("%c (%d,%d) -> (%d,%d) %d\n", c,i,j,up,j,t+1);
                }

                if ((down <= k-1) && (buff[down][j] != 'X') && (buff[down][j] != c)) {
                    if ((c == '+') && (positive_grid[down][j] != '+')) {
   			positive_grid[down][j] = '+';	
                        enqueue(q, c, down, j, t+1);
                    }
                    if ((c == '-') && (negative_grid[down][j] != '-')) {
                        negative_grid[down][j] = '-';
                        enqueue(q, c, down, j, t+1);
                    }
                    //printf("%c (%d,%d) -> (%d,%d) %d\n", c,i,j,down,j,t+1);
                }

                if ((right <= l-1) && (buff[i][right] != 'X') && (buff[i][right] != c)){
                    if ((c == '+') && (positive_grid[i][right] != '+')) {
   			positive_grid[i][right] = '+';	
                        enqueue(q, c, i, right, t+1);
                    }
                    if ((c == '-') && (negative_grid[i][right] != '-')) {
                        negative_grid[i][right] = '-';
                        enqueue(q, c, i, right, t+1);
                    }
                    //printf("%c (%d,%d) -> (%d,%d) %d\n", c,i,j,up,j,t+1);
                }

                if ((left >= 0) && (buff[i][left] != 'X') && (buff[i][left] != c)){
                    if ((c == '+') && (positive_grid[i][left] != '+')) {
   			positive_grid[i][left] = '+';	
                        enqueue(q, c, i, left, t+1);
                    }
                    if ((c == '-') && (negative_grid[i][left] != '-')) {
                        negative_grid[i][left] = '-';
                        enqueue(q, c, i, left, t+1);
                    }
                    //printf("%c (%d,%d) -> (%d,%d) %d\n", c,i,j,up,j,t+1);
                }

                //		free(elem);
            }
        }
        free(elem);
        elem = dequeue(q);
    }

    free(elem);

    if(doomsday)
        printf("%d\n", time);
    else
        printf("the world is saved\n");

    for(i=0; i<k; i++)
    {
        for(j=0; j<l; j++)
        {	
            printf("%c", buff[i][j]);
        }
        printf("\n");
    }

    fclose(fp);

    //destroy(queue);
    return 0;
} 

File: doomsday.py
#!/usr/bin/env python3

from collections import deque

def doomsday(f):

	i = j = k = l = t = doomsday = 0
	time = -1
	buff = [[0 for j in range(1000)] for i in range(1000)]
	positive_grid = [["." for j in range(1000)] for i in range(1000)]
	negative_grid = [["." for j in range(1000)] for i in range(1000)]
	c = f.read(1)
	q = 0

	class Node:
	
		def __init__(self,symbol,x,y,time):
			self.symbol = symbol
			self.x = x
			self.y = y
			self.time = time

	while (c):
		buff[i][j] = c

		if c == "+" or c == "-":
			node = Node(c, i, j, t)
			if q:
				q.append(node)
			else:
				q = deque([node])
			if c == "+":
				positive_grid[i][j] = "+"
			if c == "-":
				negative_grid[i][j] = "-"
		
		j += 1
		c = f.read(1)
		
		if j == 1000:
			i += 1
			k = i 
			l = j 
			j = 0 

		if c == "\n":
			i += 1
			k = i 
			l = j 
			j = 0 
			while c == "\n":
				c = f.read(1)
	
	up = down = right = left = 0 

	while q:
		
		elem = q.popleft()
		c = elem.symbol
		i = elem.x
		j = elem.y 
		t = elem.time

		if doomsday and time < t:
			break
	
		if (c == "-" and buff[i][j] == "+") or (c == "+" and buff[i][j] == "-"):
			buff[i][j] = "*"
			if doomsday == 0:
				doomsday = 1
				time = t
				
		elif buff[i][j] != "*":
			if time == -1 or t <= time:
				buff[i][j] = c
	
			up = i - 1
			down = i + 1
			right = j + 1
			left = j - 1
		
			if up >= 0 and buff[up][j] != "X" and buff[up][j] != c:
				if c == "+" and positive_grid[up][j] != "+":
					positive_grid[up][j] = "+"
					node = Node(c, up, j, t+1)
					q.append(node)
				if c == "-" and negative_grid[up][j] != "-":
					negative_grid[up][j] = "-"
					node = Node(c, up, j, t+1)
					q.append(node)
					
			if down <= k-1 and buff[down][j] != "X" and buff[down][j] != c:
				if c == "+" and positive_grid[down][j] != "+":
					positive_grid[down][j] = "+"
					node = Node(c, down, j, t+1)
					q.append(node)
				if c == "-" and negative_grid[down][j] != "-":
					negative_grid[down][j] = "-"
					node = Node(c, down, j, t+1)
					q.append(node)

			if right <= l-1 and buff[i][right] != "X" and buff[i][right] != c:
				if c == "+" and positive_grid[i][right] != "+":
					positive_grid[i][right] = "+"
					node = Node(c, i, right, t+1)
					q.append(node)
				if c == "-" and negative_grid[i][right] != "-":
					negative_grid[i][right] = "-"
					node = Node(c, i, right, t+1)
					q.append(node)

			if left >= 0 and buff[i][left] != "X" and buff[i][left] != c:
				if c == "+" and positive_grid[i][left] != "+":
					positive_grid[i][left] = "+"
					node = Node(c, i, left, t+1)
					q.append(node)
				if c == "-" and negative_grid[i][left] != "-":
					negative_grid[i][left] = "-"
					node = Node(c, i, left, t+1)
					q.append(node)

	if doomsday:
		print(time)
	else:
		print("the world is saved")

	for i in range(k):
		for j in range(l):
			sys.stdout.write(buff[i][j])
		print()
	
if __name__ == "__main__":
	import sys
	if len(sys.argv) > 1:
		with open(sys.argv[1], "rt") as f:
			doomsday(f)
	else:
		print("Usage: ./doomsday text_file")
